server/stream.go

```diff
@@ -597,6 +623,95 @@ func (mset *Stream) setupStore(fsCfg *FileStoreConfig) error {
 	return nil
 }
 
+// NumMsgIds returns the number of message ids being tracked for duplicate suppression.
+func (mset *Stream) NumMsgIds() int {
+	mset.mu.RLock()
+	defer mset.mu.RUnlock()
+	return len(mset.ddmap)
+}
+
+// checkMsgId will process and check for duplicates.
+// Lock should be held.
+func (mset *Stream) checkMsgId(id string) *ddentry {
+	if id == "" || mset.ddmap == nil {
+		return nil
+	}
+	return mset.ddmap[id]
+}
+
+// Will purge the entries that are past the window.
+// Should be called from a timer.
+func (mset *Stream) purgeMsgIds() {
+	mset.mu.Lock()
+	defer mset.mu.Unlock()
+
+	now := time.Now().UnixNano()
+	tmrNext := mset.config.Duplicates
+	window := int64(tmrNext)
+
+	for i, dde := range mset.ddarr[mset.ddindex:] {
+		if now-dde.ts >= window {
+			delete(mset.ddmap, dde.id)
+		} else {
+			mset.ddindex += i
+			// Check if we should garbage collect here if we are 1/3 total size.
+			if cap(mset.ddarr) > 3*(len(mset.ddarr)-mset.ddindex) {
+				mset.ddarr = append([]*ddentry(nil), mset.ddarr[mset.ddindex:]...)
+				mset.ddindex = 0
+			}
+			tmrNext = time.Duration(window - (now - dde.ts))
+			break
+		}
+	}
+	if len(mset.ddmap) > 0 {
+		// Make sure to not fire too quick
+		const minFire = 50 * time.Millisecond
+		if tmrNext < minFire {
+			tmrNext = minFire
+		}
+		mset.ddtmr.Reset(tmrNext)
+	} else {
+		mset.ddtmr.Stop()
+		mset.ddtmr = nil
+	}
+}
+
+// storeMsgId will store the message id for duplicate detection.
+func (mset *Stream) storeMsgId(dde *ddentry) {
+	mset.mu.Lock()
+	if mset.ddmap == nil {
+		mset.ddmap = make(map[string]*ddentry)
+	}
+	if mset.ddtmr == nil {
+		mset.ddtmr = time.AfterFunc(mset.config.Duplicates, mset.purgeMsgIds)
+	}
+	mset.ddmap[dde.id] = dde
+	mset.ddarr = append(mset.ddarr, dde)
+	mset.mu.Unlock()
+}
+
+// Will return the value for the header denoted by key or nil if it does not exists.
+// This function ignores errors and tries to achieve speed and no additional allocations.
+func getHdrVal(key string, hdr []byte) []byte {
+	index := bytes.Index(hdr, []byte(key))
+	if index < 0 {
+		return nil
+	}
+	var value []byte
+	for i := index + len(key) + 2; i > 0 && i < len(hdr); i++ {
+		if hdr[i] == '\r' && i < len(hdr)-1 && hdr[i+1] == '\n' {
+			break
+		}
+		value = append(value, hdr[i])
```

Comment:
Should we return a slice instead of appending bytes?

Reply:
Actually opposite, do not want or actually can reference a blind read buffer that could be large.

Reply:
Right. I was just wondering if this would be more of the caller to decide if a copy is required or not. But I guess that if a message ID is present it will always need to be stored (except in error conditions, but that's ok), so we better make the copy here.